### ISSUE TITLE
Some way of indicating XSUB would be nice

### ISSUE THREAD

**Issue by ribasushi** [kentnl/Devel-Isa-Explainer#1]
`(*)` perhaps?
An XSUB can be detected [like this](https://metacpan.org/source/RIBASUSHI/DBIx-Class-0.082821/t/lib/DBICTest/Util/LeakTracer.pm#L262-264)

**Comment by kentfredric** [comment]
Is there a specific way I should obtain references to subs for best results?

Given I'm presently just doing string mangling from `mro::` results.

**Comment by kentfredric** [comment]
well, not just that, I'm using Package::Stash ... to get symbol names. I guess I could factor that better so I can just reuse that object. :)

**Comment by ribasushi** [comment]
> Is there a specific way I should obtain references to subs for best results?

The canonical thing is `->can`. That will unambiguously return the "thing perl would call". You can also rely on `P::S`, which will return the right thing almost in all cases, unless there is a stale mro cache entry (extremely rare).

**Comment by kentfredric** [comment]
Just I'm worried that with `->can`, I'm going to wind up in a situation where Package::Stash reports "Namespace::sub" exists, but its a stale cache entry, so the data says "I have a sub", but the information resolved by `->can()` instead refers to the _parent_ classes `sub` due to `can` invoking Perl MRO lookup mechanics.

At least if I use Package::Stash consistently everywhere, I'm going to be consistent in what I'm reporting. 

That way the stale MRO cache entry will resolve as "no" for an XSUB check, but its parent will return a "Yes" for an XSUB check ( which is desirable ), instead of _both_ returning "Yes" for an XSUB check ( due to resolving to the same coderef )

**Comment by ribasushi** [comment]
Correct, this is why I recommended in the very beginning you use `->can` **throughout the entire stack**. If `Child->can("foo") == Parent->can("foo")`, then either "foo" is inherited, or it has been inserted into the symtable via `*Child::foo = \&Parent::foo`. That difference is immaterial to isa-splain.

Additionally using `->can` this way allows you to properly report the ancestry of [these weird-ass cases](http://grep.cpan.me/?q=sub%5Cs%2Bcan%5Cb).

**Comment by kentfredric** [comment]
Ah, so you mean perform secondary validation on the results of Package::Stash->get_all_symbols('CODE') and skip refs that are unmodified from their parent? 

K. Simple enough.

**Comment by kentfredric** [comment]
This logic is confusing me:

> Child->can("foo") == Parent->can("foo")

How do we determine what "Parent" is in this equation?

Do we:

a: Simply perform the same query on all `@ISA`, ie:

``` perl
     foreach my $parent ( @{$child::ISA} ) {
             if ( $child->can('sub') == $parent->can('sub')  ) { 
                   ## sub is a cached node, skip it
             }
     }
```

?

Or do we have to probe linearized isa?

**Comment by ribasushi** [comment]
Forget directly looking at the `@ISA` variable entirely (it won't give you a sensible answer on MI codebases). `mro::get_lineari_isa` is the only way.

And yes - you check each class in a hierarchy independently, and then "recombine" the result: in essence replicating what the p5 OO dispatch does.

**Comment by kentfredric** [comment]
So is the correct approach:

```
     foreach my $parent ( linearised_isa($child) ) {
             next unless $parent->can('sub');
             last if $child->can('sub') != $parent->can('sub');
             skip('sub');                   ## sub is a cached node, skip it
     }
```

Sorry, I'm extra dumb today.

**Comment by ribasushi** [comment]
Yep, you got it (though the skip + comment confuses me, but that's likely a leftover from something else)

**Comment by ribasushi** [comment]
Though note that mro::get_linear_isa($class) always returns $class at the first position, so you can skip checking that

**Comment by kentfredric** [comment]
I mean "skip" means "sub foo does not actually exist (as far as our purposes are concerned) in $child, so don't show it in child" . Though I might have been side-tracked by the whole Package::Stash debacle.

**Comment by ribasushi** [comment]
I think you are getting things confused. Let me give you a bird-eye view of what the proper way would be taking in consideration all of MI, 5.8 mro, and mro in general:

You use P::S **once** in the entire program like so:

```
~$ perl -MPackage::Stash -MDevel::Dwarn -MDBIx::Class::Core -e '
  my $all_found;
  $all_found->{$_}++ for map
    { Package::Stash->new($_)->list_all_symbols("CODE") }
    @{mro::get_linear_isa("DBIx::Class::Core")}
  ;
  Dwarn $all_found
'
```

Then you forget `P::S` exists, and use the "`->can` crawl" to actually figure out who came from where.

Does this make more sense?

**Comment by kentfredric** [comment]
Not hugely, I think my brain has dieded. 

> use the "->can crawl"

^^^ this is _still_ where I get lost.

**Comment by kentfredric** [comment]
Ok, I'll just make sure I've got the right idea with regards to "output", at least that way I know what the goal is ( because I'm still partly unclear on that! )

Consider this:

``` perl
use strict;
use warnings;

use Test::More;

# ABSTRACT: Test for isa-cache overrides being detected

use Devel::Isa::Explainer ();
use Data::Dump qw(pp);

@ClassA::ISA = ('ClassB');
@ClassB::ISA = ('ClassC');

#sub ClassB::fun { print q[world] }
sub ClassC::fun { print q[hello] }
{
  no warnings 'once';
  *ClassA::fun = \&ClassC::fun;
}
*extract_mro = \&Devel::Isa::Explainer::_extract_mro;
$INC{ 'Class' . $_ . q[.pm] } = 1 for qw( A B C );

note pp extract_mro('ClassA');

done_testing;
```

Is the output as follows what is expected?

``` perl
 [
   { class => "ClassA", subs => {} },
   { class => "ClassB", subs => {} },
   {
     class => "ClassC",
     subs  => { fun => { shadowed => 0, shadowing => 0, xsub => 0 } },
   },
]
```

And if I uncomment `#sub ClassB::fun { print q[world] }`, is the graph becomming:

``` perl
 [
   {
     class => "ClassA",
     subs  => { fun => { shadowed => 0, shadowing => 1, xsub => 0 } },
   },
   {
     class => "ClassB",
     subs  => { fun => { shadowed => 1, shadowing => 1, xsub => 0 } },
   },
   {
     class => "ClassC",
     subs  => { fun => { shadowed => 1, shadowing => 0, xsub => 0 } },
   },
 ]
```

What is expected?

**Comment by ribasushi** [comment]
I was typing up something, here is the minimal gist:

```
perl -MPackage::Stash -MDevel::Dwarn -MDBIx::Class::Core -e '
  my @isa = @{mro::get_linear_isa("DBIx::Class::Core")};

  my $all_methods = { map { $_ => [] } map
    { Package::Stash->new($_)->list_all_symbols("CODE") }
    @isa
  };


  for my $m (keys %$all_methods) {
    my $currently_visible;

    # looking up in reverse is important, as we care about "first seen at"
    for my $c (reverse @isa) {
      my $cref = $c->can($m) or next;
      if (
        ! defined $currently_visible
          or
        $cref != $currently_visible
      ) {
        # unshift because of the reverse
        unshift @{$all_methods->{$m}}, $c;
        $currently_visible = $cref;
      }
    }
  }

  Dwarn $all_methods;
'
```
